### PR TITLE
make shadowing a builtin a warning

### LINKED ISSUE #3319: Make functions shadowing builtins a warning, not a failure
#3013 added the `--shadow-builtin` flag, but I am concerned that it is setting up problems with forward compatibility. In particular, if you write a function today called `foo`, and then we introduce a builtin called `foo` down the track, your function will stop working completely.

The options here are not great - if we do nothing, your shell might explode (see the multiple `test` functions), but a soft warning rather than a hard fail is more acceptable as it is less likely to cause sudden explosions.

### PR DESCRIPTION
**Pull request by krader1961**
## Description

@zanchey and @faho, Any obvious problems with this change? It turns the `function $builtin` into a simple warning with an improved error message. Attempting to `function not_builtin --shadow-builtin` is still a hard error since we don't want people to "future proof" their functions against the introduction of a builtin by the same name.

Fixes issue #3319

## TODOs:

- [x] Changes to fish usage are reflected in user documenation/manpages.
- [x] Tests have been added for regressions fixed

Further discussion has lead to the decision that defining a function that
shadows a builtin without using `--shadow-builtin` should not keep the
function from being created. It should still print a warning to stderr
so the user knows there may be problems. This also improves the wording
of the respective error messages.

Fixes #3319

### CODE REVIEW

**Comment by floam** [comment]
> since we don't want people to "future proof" their functions against the introduction of a builtin by the same name.

Why not, really? If they think they know what they're doing, I think that's enough. It's something we might even do at some point if we knew a new builtin was coming.

**Comment by floam** [comment]
I think it'd be better to make the above case cause a `debug(1, ..)` diagnostic.

**Comment by floam** [comment]
Remember, the primary goal is to simply protect users who might accidentally, unknowingly redefine a function while goofing around with fish, from having strange stuff happen after for instance making a `function test`. Not introduce a new stricter, more explicit scheme of function definition.

**Comment by floam** [comment]
Also, it's not just mindless future-proofing where one would do that, but being able to provide a wrapper that works on both a current version of fish and older ones where we didn't implement something as a builtin yet, without resorting to hacks requiring them to do it twice, again with a hack like in our `string.fish` with just a the removal of `--shadows-builtin` from the `function` line.

**Comment by zanchey** [comment]
I agree that `--shadow-builtin` on a non-builtin should not fail; it creates compatibility problems unnecessarily.

**Comment by krader1961** [comment]
> I agree that --shadow-builtin on a non-builtin should not fail;

See my comment on issue #3319. I strongly disagree. If we're going to allow this then we should completely revert the original change.

**Comment by krader1961** [comment]
> Also, it's not just mindless future-proofing where one would do that, but being able to provide a wrapper that works on both a current version of fish and older ones where we didn't implement something as a builtin yet, without resorting to hacks requiring them to do it twice,

They absolutely have to provide two versions. Because it is extremely unlikely that their original function adequately mimics the new builtin. We do not want it to be easy for people to "future proof" (scare quotes are intentional) their function by using the `--shadow-builtin` flag when their function is used on fish versions without the builtin. Allowing that makes this feature pointless.

Someone providing a "wrapper" meant to provide functionality in old versions of fish that don't have the builtin aren't creating a wrapper function in the sense of our `history` function. If they are writing a wrapper their function will have a `builtin $func_name` statement (where "$func_name" will be the name of the function) in which case it won't work on old fish versions without the builtin. So by definition a "wrapper" function in the sense you seem to be using the term requires two implementations.

The author of such a function could distinguish between a fish version that supports or does not support the builtin in a single function rather than creating two distinct implementations. If we want to allow that then we should remove the `--shadow-builtin` flag and not emit any warnings or errors.

**Comment by krader1961** [comment]
> I think it'd be better to make the above case cause a debug(1, ..) diagnostic.

I agree. The reasons I didn't do so is 1) I was emulating the existing code and 2) the whole "debug(0," or "debug(1," idiom to emit an error or warning rather than a debug message is FUBAR so I tend not to think about using it. Despite my work on improving the utility of the actual debug messages.

File: src/builtin.cpp [hunk 1]
```diff
@@ -88,7 +88,8 @@ bool builtin_data_t::operator<(const builtin_data_t *other) const {
 ///
 int builtin_count_args(const wchar_t *const *argv) {
     int argc;
-    for (argc = 1; argv[argc] != NULL; argc++);
+    for (argc = 1; argv[argc] != NULL; argc++)
+        ;
```
Comment (floam) [inline src/builtin.cpp]: Why does clang-format do this? Is that really the default behavior or have we messed up some options to it? It should create braces for us or just leave it alone.

**Comment by zanchey** [comment]
> The author of such a function could distinguish between a fish version that supports or does not support the builtin in a single function rather than creating two distinct implementations.

This was the case I was concerned about - trying to use a builtin if it is available, and falling back if not - but as you say, `contains somebuiltin (builtin -n)` should do the job.

File: doc_src/function.txt [hunk 2]
```diff
@@ -29,7 +29,7 @@ The following options are available:
...
+- `-B` or `--shadow-builtin` should be specified if the function name is the same as a builtin. Specifying this flag indicates your acknowledgement that you are wrapping or replacing the builtin command. This is a safety feature to make it harder for people to inadvertently break the shell by doing things like `function test; return 0; end`. If you do that you'll get a warning message written to stderr but your function will still be created. If the function name is not currently a builtin using this flag will produce an error and the function will not be created. If you want to write a function that provides a builtin to an older version of fish you need to add something like `builtin --names | grep -q '^cmd$'; and return` to the top of the function script (where `cmd` is the name of the builtin/function). That will keep your script from replacing the builtin with your function on the newer fish version while allowing your function to provide similar functionality on older versions of fish.
```
Comment (zanchey) [inline doc_src/function.txt]: I would reword this text as follows:

`-B` or `--shadow-builtin` silences the warning that is produced when the function name is the same as a builtin. It should be specified when replacing or wrapping a builtin command. This is a safety feature to maintain the integrity of the environment; a common mistake is to create a function called `test`, when much of the shell internals rely on the behaviour of the existing test builtin. Without this flag, a warning message will be printed, although the function will still be created. Note that, if the function name is not builtin, this flag will produce an error and the function will not be created. To create fallback functions for old versions of fish, consider using a test like `contains function_name (builtin --names)` before defining the function.

File: src/builtin.cpp [hunk 3]
```diff
@@ -1731,15 +1732,13 @@ int builtin_function(parser_t &parser, io_streams_t &streams, const wcstring_lis
                 }
             }
             if (function_name_shadows_builtin && !shadow_builtin) {
-                append_format(
-                    *out_err,
-                    _(L"%ls: function name shadows a builtin so you must use '--shadow-builtin'"),
-                    argv[0]);
-                res = STATUS_BUILTIN_ERROR;
+                debug(1, _(L"%ls \"%ls\" shadows a builtin so you should use '--shadow-builtin' "
```
Comment (zanchey) [inline src/builtin.cpp]: I think a better phrase is "a builtin command "%ls" already exists (use --shadow-builtin to suppress this warning)"

**Comment by krader1961** [comment]
I'm going to simply revert the original change since it is clear no one likes it now even though at the time it was committed no one objected.
